stdlib/bytesLabels.mli

```diff
@@ -302,6 +302,22 @@ val equal: t -> t -> bool
 (** The equality function for byte sequences.
     @since 4.05.0 *)
 
+val to_hex : bytes -> bytes
+(** Return the printable hexadecimal representation of the given byte sequence.
+
+    @since 4.11.0
+*)
+
+val from_hex : bytes -> bytes
+(** Convert a hexadecimal representation back into the corresponding byte
+    sequence.
+
+    @raise Invalid_argument if the argument is not an even number of hexadecimal
```

Comment:
This is a parsing function. A parsing function is expected to fail, it should not raise `Invalid_argument`, at most `Failure` but I propose [this signature](https://erratique.ch/software/b0/doc/B00_std/String/Ascii/index.html#val-of_hex) instead, which is more useful.

Reply:
I don't agree - `parse_hex` is a parsing function; `from_hex` is a conversion function. I agree that `Failure` is more appropriate than `Invalid_argument`, but I'm not convinced that it would be sensible to have `Digest.from_hex` raise `Invalid_argument` with these new ones raising `Failure`.

Quite happy to add `parse_hex`/`from_hex_opt` if wanted (my original purpose was this was to expose a function we already had)

Reply:
I don't think the distinction you are trying to make is useful in practice. Basically if you read hex from a string to convert it into bytes you should expect it may fail.

Besides if new functions are added to the stdlib it would be good to at least make them *useful*. What you are proposing is terrible for error reporting and basically entails that if one wants to do good error reporting you will have to rewrite the functionality.

Reply:
If I've just lexed a hexadecimal string, then there's no way that converting it is going to fail, and error handling is a bit-rotting waste of effort - the failure mode has already been dealt with by the lexer. If the only API is the `option`/`result` returning one, then there's at least a pointless `match` or call to `Option.get`, `Result.get_ok` with the same overall effect that if there's a bizarre programmer error (like I suddenly start parsing `z` as a valid hex character) then I still get `Invalid_argument` just from a different function. There's usually a case for both functions, but if you're only writing one then it should be the exception raising version.

Reply:
> If I've just lexed a hexadecimal string, then there's no way that converting it is going to fail, and error handling is a bit-rotting waste of effort

I'm not sure I understand you if you just parsed an hexadecimal representation to bytes (`of_hex`) then the conversion back to an hexadecimal representation (`to_hex`) cannot fail.

If you meant that once you converted to hexadecimal (`to_hex`) you want to immediately get back to bytes (`of_hex`) then I argue that's not what you are going to do in 99% of uses of the function. 

For the remaining 1% composing with `Result.get_ok` is entirely sufficient and effortless. There's no need to clutter the API with multiple versions of the same function.

Reply:
What @dra27 means is: if you use an `ocamllex`-generated lexer to input an even-length string of hexadecimal digits, then pass it to `of_hex`, you know for sure that it won't fail so the call to `Result.get_ok` is just noise.